### PR TITLE
Add SHOW VITESS_REPLICATION_STATUS Query Support

### PR DESCRIPTION
**Pull request by mattlord**
### What
The new `SHOW VITESS_REPLICATION_STATUS` command/query shows the MySQL replica/replication (_not_ `vreplication`) health for the vitess deployment. It returns a row for each `REPLICA` and `RDONLY` tablet in the topology -- with support for filtering by `Keyspace/Shard` using a `LIKE` clause -- providing relevant health and status information, including the current [throttler status for the tablet](https://vitess.io/docs/reference/features/tablet-throttler/).

> Note: This is **_experimental_** for 12.0 as I would like to have additional flexibility to tweak this based on user feedback over the coming months.

### Why
This allows applications -- which are generally interfacing with Vitess through a VTGate -- to alter behavior based on the replication state. They can e.g. decide to back off or delay some planned batch work if replication lag is high or if the [tablet throttler](https://vitess.io/docs/reference/features/tablet-throttler/) is configured and shows that there are too many requests coming in.

This also offers humans and other tooling an easy and convenient way to examine the replication health and related status of the entire Vitess topology, a Keyspace, or a Shard.

### Examples
```
mysql> show vitess_tablets;
+-------+----------+-------+------------+---------+------------------+--------------+----------------------+
| Cell  | Keyspace | Shard | TabletType | State   | Alias            | Hostname     | PrimaryTermStartTime |
+-------+----------+-------+------------+---------+------------------+--------------+----------------------+
| zone1 | commerce | 0     | PRIMARY    | SERVING | zone1-0000000100 | 7e2718a5ab58 | 2021-10-03T20:05:04Z |
| zone1 | commerce | 0     | REPLICA    | SERVING | zone1-0000000101 | 7e2718a5ab58 |                      |
| zone1 | commerce | 0     | RDONLY     | SERVING | zone1-0000000102 | 7e2718a5ab58 |                      |
| zone1 | customer | -80   | PRIMARY    | SERVING | zone1-0000000300 | 7e2718a5ab58 | 2021-10-03T20:08:30Z |
| zone1 | customer | -80   | REPLICA    | SERVING | zone1-0000000301 | 7e2718a5ab58 |                      |
| zone1 | customer | -80   | RDONLY     | SERVING | zone1-0000000302 | 7e2718a5ab58 |                      |
| zone1 | customer | 80-   | PRIMARY    | SERVING | zone1-0000000400 | 7e2718a5ab58 | 2021-10-03T20:08:30Z |
| zone1 | customer | 80-   | REPLICA    | SERVING | zone1-0000000401 | 7e2718a5ab58 |                      |
| zone1 | customer | 80-   | RDONLY     | SERVING | zone1-0000000402 | 7e2718a5ab58 |                      |
+-------+----------+-------+------------+---------+------------------+--------------+----------------------+
9 rows in set (0.00 sec)

mysql> show vitess_replication_status;
+----------+-------+------------+------------------+--------------+--------------------+-------------------------------------------------------------------------+----------------+-----------------------------------------+
| Keyspace | Shard | TabletType | Alias            | Hostname     | ReplicationSource  | ReplicationHealth                                                       | ReplicationLag | ThrottlerStatus                         |
+----------+-------+------------+------------------+--------------+--------------------+-------------------------------------------------------------------------+----------------+-----------------------------------------+
| commerce | 0     | REPLICA    | zone1-0000000101 | 7e2718a5ab58 | 7e2718a5ab58:17100 | {"EventStreamRunning":"Yes","EventApplierRunning":"Yes","LastError":""} | 0              | {"state":"OK","load":0.00,"message":""} |
| commerce | 0     | RDONLY     | zone1-0000000102 | 7e2718a5ab58 | 7e2718a5ab58:17100 | {"EventStreamRunning":"Yes","EventApplierRunning":"Yes","LastError":""} | 0              | {"state":"OK","load":0.00,"message":""} |
| customer | -80   | PRIMARY    | zone1-0000000300 | 7e2718a5ab58 | :0                 | {"EventStreamRunning":"","EventApplierRunning":"","LastError":""}       | 0              | {"state":"OK","load":0.00,"message":""} |
| customer | -80   | REPLICA    | zone1-0000000301 | 7e2718a5ab58 | 7e2718a5ab58:17300 | {"EventStreamRunning":"Yes","EventApplierRunning":"Yes","LastError":""} | 0              | {"state":"OK","load":0.00,"message":""} |
| customer | -80   | RDONLY     | zone1-0000000302 | 7e2718a5ab58 | 7e2718a5ab58:17300 | {"EventStreamRunning":"Yes","EventApplierRunning":"Yes","LastError":""} | 0              | {"state":"OK","load":0.00,"message":""} |
| customer | 80-   | REPLICA    | zone1-0000000401 | 7e2718a5ab58 | 7e2718a5ab58:17400 | {"EventStreamRunning":"Yes","EventApplierRunning":"Yes","LastError":""} | 0              | {"state":"OK","load":0.00,"message":""} |
| customer | 80-   | RDONLY     | zone1-0000000402 | 7e2718a5ab58 | 7e2718a5ab58:17400 | {"EventStreamRunning":"Yes","EventApplierRunning":"Yes","LastError":""} | 0              | {"state":"OK","load":0.00,"message":""} |
+----------+-------+------------+------------------+--------------+--------------------+-------------------------------------------------------------------------+----------------+-----------------------------------------+
7 rows in set (0.01 sec)

mysql> show vitess_replication_status like "customer/-80";
+----------+-------+------------+------------------+--------------+--------------------+-------------------------------------------------------------------------+----------------+-----------------------------------------+
| Keyspace | Shard | TabletType | Alias            | Hostname     | ReplicationSource  | ReplicationHealth                                                       | ReplicationLag | ThrottlerStatus                         |
+----------+-------+------------+------------------+--------------+--------------------+-------------------------------------------------------------------------+----------------+-----------------------------------------+
| customer | -80   | PRIMARY    | zone1-0000000300 | 7e2718a5ab58 | :0                 | {"EventStreamRunning":"","EventApplierRunning":"","LastError":""}       | 0              | {"state":"OK","load":0.00,"message":""} |
| customer | -80   | REPLICA    | zone1-0000000301 | 7e2718a5ab58 | 7e2718a5ab58:17300 | {"EventStreamRunning":"Yes","EventApplierRunning":"Yes","LastError":""} | 0              | {"state":"OK","load":0.00,"message":""} |
| customer | -80   | RDONLY     | zone1-0000000302 | 7e2718a5ab58 | 7e2718a5ab58:17300 | {"EventStreamRunning":"Yes","EventApplierRunning":"Yes","LastError":""} | 0              | {"state":"OK","load":0.00,"message":""} |
+----------+-------+------------+------------------+--------------+--------------------+-------------------------------------------------------------------------+----------------+-----------------------------------------+
3 rows in set (0.01 sec)

mysql> show vitess_replication_status like "commerce%";
+----------+-------+------------+------------------+--------------+--------------------+-------------------------------------------------------------------------+----------------+-----------------------------------------+
| Keyspace | Shard | TabletType | Alias            | Hostname     | ReplicationSource  | ReplicationHealth                                                       | ReplicationLag | ThrottlerStatus                         |
+----------+-------+------------+------------------+--------------+--------------------+-------------------------------------------------------------------------+----------------+-----------------------------------------+
| commerce | 0     | REPLICA    | zone1-0000000101 | 7e2718a5ab58 | 7e2718a5ab58:17100 | {"EventStreamRunning":"Yes","EventApplierRunning":"Yes","LastError":""} | 0              | {"state":"OK","load":0.00,"message":""} |
| commerce | 0     | RDONLY     | zone1-0000000102 | 7e2718a5ab58 | 7e2718a5ab58:17100 | {"EventStreamRunning":"Yes","EventApplierRunning":"Yes","LastError":""} | 0              | {"state":"OK","load":0.00,"message":""} |
+----------+-------+------------+------------------+--------------+--------------------+-------------------------------------------------------------------------+----------------+-----------------------------------------+
2 rows in set (0.00 sec)

$ curl -s "localhost:15101/throttler/check?app=vtgate"
{"StatusCode":200,"Value":0,"Threshold":0,"Message":""}

##
# Perform a direct insert on the commerce REPLICA and then perform the same INSERT on the primary via vtgate
##

mysql> insert into commerce.customer (email) values ("mlord@planetscale.com");
Query OK, 1 row affected (0.01 sec)

mysql> show vitess_replication_status like "commerce%"\G
*************************** 1. row ***************************
         Keyspace: commerce
            Shard: 0
       TabletType: REPLICA
            Alias: zone1-0000000101
         Hostname: 7e2718a5ab58
ReplicationSource: 7e2718a5ab58:17100
ReplicationHealth: {"EventStreamRunning":"Yes","EventApplierRunning":"No","LastError":"Could not execute Write_rows event on table vt_commerce.customer; Duplicate entry '1' for key 'PRIMARY', Error_code: 1062; handler error HA_ERR_FOUND_DUPP_KEY; the event's master log vt-0000000100-bin.000001, end_log_pos 9646"}
   ReplicationLag: 15
  ThrottlerStatus: {"state":"OK","load":0.00,"message":""}
*************************** 2. row ***************************
         Keyspace: commerce
            Shard: 0
       TabletType: RDONLY
            Alias: zone1-0000000102
         Hostname: 7e2718a5ab58
ReplicationSource: 7e2718a5ab58:17100
ReplicationHealth: {"EventStreamRunning":"Yes","EventApplierRunning":"Yes","LastError":""}
   ReplicationLag: 0
  ThrottlerStatus: {"state":"OK","load":0.00,"message":""}
2 rows in set (0.01 sec)
```

Signed-off-by: Matt Lord <mattalord@gmail.com>

## Related Issue(s)
https://github.com/vitessio/vitess/issues/6984

## Checklist
- [ ] Should this PR be backported?
- [x] Tests were added or are not required
- [x] Documentation was added: https://github.com/vitessio/website/pull/839

## Deployment Notes
I'd like to flag this as `experimental` so that we can get it into 12.0 but have some added flexibility in modifying/improving it based on user feedback.

### CODE REVIEW

File: go/vt/sqlparser/ast_format.go [hunk 1]
```diff
@@ -868,9 +869,6 @@ func (node *AliasedTableExpr) Format(buf *TrackedBuffer) {
 	buf.astPrintf(node, "%v%v", node.Expr, node.Partitions)
 	if !node.As.IsEmpty() {
 		buf.astPrintf(node, " as %v", node.As)
-		if len(node.Columns) != 0 {
```
Comment (systay) [inline go/vt/sqlparser/ast_format.go:871]: Why do you think we should remove this?

Comment (mattlord) [reply go/vt/sqlparser/ast_format.go:871]: Didn't intend to -- must have been a merge issue. Re-adding it. Thanks!

**Review by sougou — APPROVED** [review]
I think this is a good start. We should eventually add more info, like running state, error, etc. There is some overlap with the experimental `VExec` command, but I think that one didn't quite take off like we expected.

Tagging @systay @harshit-gangal to look at the parser side.

Optimistically approving.

File: go/vt/vtgate/executor.go [hunk 2]
```diff
@@ -63,6 +64,39 @@ import (
 // this is the healthcheck used by vtgate, used by the "vstream * from" functionality
 var vtgateHealthCheck discovery.HealthCheck
 
+var getTabletThrottlerStatus = func(tabletAddr string) (string, error) {
```
Comment (systay) [inline go/vt/vtgate/executor.go:67]: I don't get it. Why isn't this just a normal function instead of being a function variable?

Comment (mattlord) [reply go/vt/vtgate/executor.go:67]: Just noting that we discussed this via Slack and I will be making it a normal function. I wasn't sure why it was used here so just aped it originally: https://github.com/vitessio/vitess/blob/59031fd32f3e3333b1263b897bb9fc0bc8e7de74/go/vt/wrangler/version.go#L36

But given that function variables generally make sense when the function is ... variable (can be overridden/changed by user flags, actions, etc) it doesn't make sense to use it here. Thanks!

File: go/vt/vtgate/executor.go [hunk 3]
```diff
@@ -991,6 +994,119 @@ func (e *Executor) showTablets(show *sqlparser.ShowLegacy) (*sqltypes.Result, er
...
+	defer cancel()
+	rows := [][]sqltypes.Value{}
+
+	// This is only used for tests
+	if UsingLegacyGateway() {
+		status := e.scatterConn.GetLegacyHealthCheckCacheStatus()
+
+		for _, s := range status {
+			for _, ts := range s.TabletsStats {
+				if ts.Tablet.Type != topodatapb.TabletType_REPLICA {
```
Comment (deepthi) [inline go/vt/vtgate/executor.go:1008]: Is it intentional to limit this to `REPLICA` types? `RDONLY` tablets also replicate from primary.

Comment (mattlord) [reply go/vt/vtgate/executor.go:1008]: It was intentional but I wasn’t sure if it was the right thing to do either. 🙂 Would you prefer rdonly too with another column showing the tablet type?

**Review by systay — APPROVED** [review]

File: go/vt/vtgate/executor.go (hunk 3 above)
Comment (deepthi) [reply go/vt/vtgate/executor.go:1008]: Yes I think that will be helpful.

Comment (mattlord) [reply go/vt/vtgate/executor.go]: OK, thanks! I made this change.

**Review by deepthi — APPROVED** [review]
This is a nice enhancement. Thank you!

**Pull request merged**